File: dart_sample/lib/demos/general/offstage_demo_page.dart
import 'package:flutter/material.dart';

import '../../counter_widgets.dart';

class OffstageDemoPage extends StatefulWidget {
  const OffstageDemoPage({super.key});

  @override
  State<OffstageDemoPage> createState() => _OffstageDemoPageState();
}

class _OffstageDemoPageState extends State<OffstageDemoPage> {
  bool _offstage = true;
  bool _visible = true;

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 10,
        children: <Widget>[
          const Text(
            'Visibility + SliverVisibility + Offstage',
            style: TextStyle(fontSize: 20, color: Colors.black),
          ),
          const Text(
            'Compare replacement, maintained layout space, and layout-without-paint behavior.',
            style: TextStyle(fontSize: 14, color: Colors.black54),
          ),
          Row(
            spacing: 8,
            children: <Widget>[
              _buildButton(
                label: 'visible=true',
                onTap: () => _setVisible(true),
                width: 104,
                background: const Color(0xFFDCE3ED),
              ),
              _buildButton(
                label: 'visible=false',
                onTap: () => _setVisible(false),
                width: 110,
                background: const Color(0xFFDCE3ED),
              ),
            ],
          ),
          Text(
            'state: visible=${_visible ? 'true' : 'false'}',
            style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
          ),
          const Text(
            'maintainState=true keeps the indicator mounted; TickerMode pauses its frame callbacks while hidden.',
            style: TextStyle(fontSize: 11, color: Colors.black54),
          ),
          Visibility(
            visible: _visible,
            maintainState: true,
            child: const SizedBox(
              height: 18,
              child: LinearProgressIndicator(),
            ),
          ),
          Container(
            height: 82,
            color: const Color(0xFFF6F8FB),
            padding: const EdgeInsets.all(8),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              spacing: 8,
              children: <Widget>[
                _buildMarker('L', const Color(0xFF1D3557)),
                Visibility.maintain(
                  visible: _visible,
                  child: Container(
                    width: 88,
                    height: 42,
                    color: const Color(0xFFA8DADC),
                    child: const Center(
                      child: Text(
                        'keeps size',
                        style: TextStyle(fontSize: 11, color: Colors.black),
                      ),
                    ),
                  ),
                ),
                _buildMarker('R', const Color(0xFF457B9D)),
              ],
            ),
          ),
          Visibility(
            visible: _visible,
            replacement: Container(
              height: 42,
              color: const Color(0xFFFFE8CC),
              child: const Center(
                child: Text(
                  'Visibility replacement',
                  style: TextStyle(fontSize: 11, color: Colors.black),
                ),
              ),
            ),
            child: Container(
              height: 42,
              color: const Color(0xFFD8F3DC),
              child: const Center(
                child: Text(
                  'Visibility child',
                  style: TextStyle(fontSize: 11, color: Colors.black),
                ),
              ),
            ),
          ),
          SizedBox(
            height: 150,
            child: CustomScrollView(
              slivers: <Widget>[
                SliverToBoxAdapter(
                  child: Container(
                    height: 42,
                    color: const Color(0xFFE9ECEF),
                    child: const Center(
                      child: Text(
                        'sliver before',
                        style: TextStyle(fontSize: 11, color: Colors.black),
                      ),
                    ),
                  ),
                ),
                SliverVisibility(
                  visible: _visible,
                  replacementSliver: SliverToBoxAdapter(
                    child: Container(
                      height: 42,
                      color: const Color(0xFFFFE8CC),
                      child: const Center(
                        child: Text(
                          'replacement sliver',
                          style: TextStyle(fontSize: 11, color: Colors.black),
                        ),
                      ),
                    ),
                  ),
                  sliver: SliverToBoxAdapter(
                    child: Container(
                      height: 42,
                      color: const Color(0xFFBDE0FE),
                      child: const Center(
                        child: Text(
                          'SliverVisibility child',
                          style: TextStyle(fontSize: 11, color: Colors.black),
                        ),
                      ),
                    ),
                  ),
                ),
                SliverToBoxAdapter(
                  child: Container(
                    height: 42,
                    color: const Color(0xFFE9ECEF),
                    child: const Center(
                      child: Text(
                        'sliver after',
                        style: TextStyle(fontSize: 11, color: Colors.black),
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
          const Text(
            'When offstage=true, child is laid out but not painted/hit-tested and takes no room in parent layout.',
            style: TextStyle(fontSize: 14, color: Colors.black54),
          ),
          Row(
            spacing: 8,
            children: <Widget>[
              _buildButton(
                label: 'offstage=true',
                onTap: () => _setOffstage(true),
                width: 112,
                background: const Color(0xFFDCE3ED),
              ),
              _buildButton(
                label: 'offstage=false',
                onTap: () => _setOffstage(false),
                width: 118,
                background: const Color(0xFFDCE3ED),
              ),
            ],
          ),
          Text(
            'state: offstage=${_offstage ? 'true' : 'false'}',
            style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
          ),
          Container(
            width: 260,
            height: 190,
            color: const Color(0xFFE7EDF6),
            padding: const EdgeInsets.all(10),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              spacing: 8,
              children: <Widget>[
                const Text(
                  'Row layout (middle child disappears from layout when offstage=true)',
                  style: TextStyle(fontSize: 11, color: Colors.black54),
                ),
                Container(
                  height: 72,
                  color: Colors.white,
                  padding: const EdgeInsets.fromLTRB(8, 10, 8, 10),
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    spacing: 8,
                    children: <Widget>[
                      _buildMarker('L', const Color(0xFF1D3557)),
                      Offstage(
                        offstage: _offstage,
                        child: Container(
                          width: 120,
                          height: 44,
                          decoration: BoxDecoration(
                            color: const Color(0xFFCCE3FF),
                            border: Border.all(
                              color: const Color(0xFF1D3557),
                              width: 2,
                            ),
                            borderRadius: BorderRadius.circular(10),
                          ),
                          child: const Center(
                            child: Text(
                              'Offstage child',
                              style: TextStyle(
                                fontSize: 11,
                                color: Colors.black,
                              ),
                            ),
                          ),
                        ),
                      ),
                      _buildMarker('R', const Color(0xFF457B9D)),
                    ],
                  ),
                ),
                const Text(
                  'Tip: switch state and watch L/R gap change.',
                  style: TextStyle(fontSize: 11, color: Colors.black54),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  static Widget _buildMarker(String label, Color color) {
    return Container(
      width: 34,
      height: 34,
      color: color,
      child: Center(
        child: Text(
          label,
          style: const TextStyle(fontSize: 12, color: Colors.white),
        ),
      ),
    );
  }

  void _setOffstage(bool value) {
    setState(() {
      _offstage = value;
    });
  }

  void _setVisible(bool value) {
    setState(() {
      _visible = value;
    });
  }
}

File: src/Sample/Plumix.Sample/Demos/General/OffstageDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/offstage_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class OffstageDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new OffstageDemoPageState();
    }
}

internal sealed class OffstageDemoPageState : State
{
    private bool _offstage = true;
    private bool _visible = true;

    public override Widget Build(BuildContext context)
    {
        return new SingleChildScrollView(
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 10,
                children:
                [
                    new Text("Visibility + SliverVisibility + Offstage", fontSize: 20, color: Colors.Black),
                    new Text(
                        "Compare replacement, maintained layout space, and layout-without-paint behavior.",
                        fontSize: 14,
                        color: Colors.DimGray),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildButton("visible=true", () => SetVisible(true), width: 104, colorHex: "#FFDCE3ED"),
                            BuildButton(
                                "visible=false",
                                () => SetVisible(false),
                                width: 110,
                                colorHex: "#FFDCE3ED"),
                        ]),
                    new Text(
                        $"state: visible={(_visible ? "true" : "false")}",
                        fontSize: 12,
                        color: Colors.DarkSlateGray),
                    new Text(
                        "maintainState=true keeps the indicator mounted; TickerMode pauses its frame callbacks "
                        + "while hidden.",
                        fontSize: 11,
                        color: Colors.DimGray),
                    new Visibility(
                        visible: _visible,
                        maintainState: true,
                        child: new SizedBox(
                            height: 18,
                            child: new LinearProgressIndicator())),
                    new Container(
                        height: 82,
                        color: Color.Parse("#FFF6F8FB"),
                        padding: new Thickness(8),
                        child: new Row(
                            mainAxisAlignment: MainAxisAlignment.Center,
                            spacing: 8,
                            children:
                            [
                                BuildMarker("L", "#FF1D3557"),
                                Visibility.Maintain(
                                    visible: _visible,
                                    child: new Container(
                                        width: 88,
                                        height: 42,
                                        color: Color.Parse("#FFA8DADC"),
                                        child: new Center(
                                            child: new Text("keeps size", fontSize: 11, color: Colors.Black)))),
                                BuildMarker("R", "#FF457B9D"),
                            ])),
                    new Visibility(
                        visible: _visible,
                        replacement: new Container(
                            height: 42,
                            color: Color.Parse("#FFFFE8CC"),
                            child: new Center(
                                child: new Text("Visibility replacement", fontSize: 11, color: Colors.Black))),
                        child: new Container(
                            height: 42,
                            color: Color.Parse("#FFD8F3DC"),
                            child: new Center(
                                child: new Text("Visibility child", fontSize: 11, color: Colors.Black)))),
                    new SizedBox(
                        height: 150,
                        child: new CustomScrollView(
                            slivers:
                            [
                                new SliverToBoxAdapter(
                                    new Container(
                                        height: 42,
                                        color: Color.Parse("#FFE9ECEF"),
                                        child: new Center(
                                            child: new Text("sliver before", fontSize: 11, color: Colors.Black)))),
                                new SliverVisibility(
                                    visible: _visible,
                                    replacementSliver: new SliverToBoxAdapter(
                                        new Container(
                                            height: 42,
                                            color: Color.Parse("#FFFFE8CC"),
                                            child: new Center(
                                                child: new Text(
                                                    "replacement sliver",
                                                    fontSize: 11,
                                                    color: Colors.Black)))),
                                    sliver: new SliverToBoxAdapter(
                                        new Container(
                                            height: 42,
                                            color: Color.Parse("#FFBDE0FE"),
                                            child: new Center(
                                                child: new Text(
                                                    "SliverVisibility child",
                                                    fontSize: 11,
                                                    color: Colors.Black))))),
                                new SliverToBoxAdapter(
                                    new Container(
                                        height: 42,
                                        color: Color.Parse("#FFE9ECEF"),
                                        child: new Center(
                                            child: new Text("sliver after", fontSize: 11, color: Colors.Black)))),
                            ])),
                new Text(
                    "When offstage=true, child is laid out but not painted/hit-tested and takes no room in parent "
                    + "layout.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("offstage=true", () => SetOffstage(true), width: 112, colorHex: "#FFDCE3ED"),
                        BuildButton("offstage=false", () => SetOffstage(false), width: 118, colorHex: "#FFDCE3ED"),
                    ]),
                new Text(
                    $"state: offstage={(_offstage ? "true" : "false")}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 260,
                    height: 190,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(10),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text(
                                "Row layout (middle child disappears from layout when offstage=true)",
                                fontSize: 11,
                                color: Colors.DimGray),
                            new Container(
                                height: 72,
                                color: Colors.White,
                                padding: new Thickness(8, 10, 8, 10),
                                child: new Row(
                                    mainAxisAlignment: MainAxisAlignment.Center,
                                    spacing: 8,
                                    children:
                                    [
                                        BuildMarker("L", "#FF1D3557"),
                                        new Offstage(
                                            offstage: _offstage,
                                            child: new Container(
                                                width: 120,
                                                height: 44,
                                                decoration: new BoxDecoration(
                                                    Color: Color.Parse("#FFCCE3FF"),
                                                    Border: Plumix.Rendering.Border.FromBorderSide(
                                                        new BorderSide(Color.Parse("#FF1D3557"), 2)),
                                                    BorderRadius: BorderRadius.Circular(10)),
                                                child: new Center(
                                                    child: new Text(
                                                        "Offstage child",
                                                        fontSize: 11,
                                                        color: Colors.Black)))),
                                        BuildMarker("R", "#FF457B9D"),
                                    ])),
                            new Text(
                                "Tip: switch state and watch L/R gap change.",
                                fontSize: 11,
                                color: Colors.DimGray),
                        ])),
                ]));
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private static Widget BuildMarker(string label, string colorHex)
    {
        return new Container(
            width: 34,
            height: 34,
            color: Color.Parse(colorHex),
            child: new Center(
                child: new Text(label, fontSize: 12, color: Colors.White)));
    }

    private void SetOffstage(bool value)
    {
        SetState(() => _offstage = value);
    }

    private void SetVisible(bool value)
    {
        SetState(() => _visible = value);
    }
}
